### ISSUE TITLE
Discussion: Do we need a UI/UX guide for developers?

### ISSUE THREAD

**Issue by pnbecker** [DSpace/dspace-angular#3211]
Many people are developing for DSpace which is great. Nevertheless, we want that DSpace look and feel like one application. We want to avoid situations where one feature looks totally different from other features or behave completely different (UX). For example we use the semantic colors from bootstrap (primary, warning, danger, success) for buttons, panels and other elements. Other useful information would be if there is a minimum width we try to optimize DSpace for. And for sure, there are many more information we should provide.

I would appreciate to hear from others, what they think: should we come up with a UI/UX guide for DSpace? What information should it contain? What unwritten rules do we have currently and what rules are we maybe missing?

**Comment by mwoodiupui** [comment]
YES.

One thing it should include is recommendations for accessibility.  The v7 UI is pleasingly colorful yet had many cases which were distinct by color but indistinct by value, making use difficult for people with certain types of vision.  And I've had to repair a few places where e.g. a control was structurally a link but styled to look like a button, which is confusing when using a screen reader.  Perhaps the guide for these conditions is simply "it is important to test with an accessibility tool and fix what it reports."

Personally I am not a very good UI designer, and I would appreciate reasonable limits on my freedom to make something ugly and hard to use.  A good style guide leaves me with fewer things to think about, because those decisions are already made.

**Comment by tdonohue** [comment]
@pnbecker and @mwoodiupui : We have a **very basic** start to this guide in the documentation at https://wiki.lyrasis.org/pages/viewpage.action?pageId=313851976
(This was begun during the 7.0 development process in a [separate page at the time](https://wiki.lyrasis.org/display/DSPACE/DSpace+UI+Design+principles+and+guidelines), but has been moved to official docs)

It's not complete or perfect **because** we do not have a UI designer available to us to write a more thorough guide.  

I'll also admit, it's **very difficult to enforce**, because enforcement has to be done manually (all reviewers/testers must be aware of the guide and do a manual check...and as we know we don't have enough reviewers).  That said, it's a starting point, and we can reference it to report bugs for pages we find that don't align well with the guide.

Overall, I agree with this idea.  I'm just pointing out the difficulties we've already faced with building such a guide.

**Comment by hieramsee** [comment]
In my opinion, the question in the headline skips an important point: the creation of a consistent UI/UX design. Only then can a guide be developed. This guide would benefit all users through the consistent further development of the DSpace UI, and as a living style guide with a component library, it could also greatly facilitate the work of developers.

DSpace currently uses Bootstrap, which is a design system with general CSS frameworks and JavaScripts for common components, but it is not a comprehensive design system for all the requirements of a complex web application with specific needs (such as integrating the corporate design of various institutions into the DSpace UI).

Ideally, UI/UX designers and developers work closely together from the beginning. Designers take care of the design of individual components, the overall look of the application, as well as user scenarios and navigation structures. The design develops gradually with design drafts and prototypes/user tests before the technical implementation begins.

In reality, however, it is often the other way around: technical developers have already developed an application and designed it themselves until it reaches a certain level of complexity. Then they realize that usability suffers and only then do they bring in UI/UX design experts. (With the topic of accessibility it is similar.)

For an already advanced application, the question arises: Do you start the optimization with solution proposals (commissioning a UI designer) or with problem analysis (user tests)? For DSpace, I would recommend starting with the fundamental optimization of the design system.

The reason for this recommendation is that in this point UX design is similar to programming: one error usually leads to many other error messages. Therefore, it makes sense to first eliminate the underlying errors. This automatically resolves many subsequent errors. Similarly, in UX design: if users do not understand the meaning and operation of single components, such as input fields that require special handling, it can lead to further misoperations throughout an entire workflow. 

Therefore, I would recommend first developing a consistent design system and then (or in parallel) creating a living style guide.
